contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.20;
+
+import "forge-std/console2.sol";
+import "../libraries/SwapIntention.sol";
+import {IQuoter} from "../interfaces/IQuoter.sol";
+import {PoolTicksCounter} from "../libraries/PoolTicksCounter.sol";
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {SafeCast} from "@uniswap/v4-core/contracts/libraries/SafeCast.sol";
+import {Pool} from "@uniswap/v4-core/contracts/libraries/Pool.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {PoolManager} from "@uniswap/v4-core/contracts/PoolManager.sol";
+
+contract Quoter is IQuoter {
+    using PoolIdLibrary for PoolKey;
+    using SafeCast for *;
+    using Hooks for IHooks;
+
+    // v4 Singleton contract
+    IPoolManager poolManager;
+
+    constructor(address _poolManager) {
+        poolManager = IPoolManager(_poolManager);
+    }
+
+    function fillSlot0(PoolId id) private view returns (Pool.Slot0 memory slot0) {
+        //TODO: extsload when storage is stable?
+        (slot0.sqrtPriceX96, slot0.tick,,) = poolManager.getSlot0(id);
+
+        return slot0;
+    }
+
+    function validateRevertReason(bytes memory reason) private pure returns (bytes memory) {
+        if (reason.length < 96) {
+            // function selector + length of bytes as uint256 + min length of revert reason padded to multiple of 32 bytes
+            if (reason.length < 68) {
+                revert UnexpectedRevertBytes();
+            }
+            assembly {
+                reason := add(reason, 0x04)
+            }
+            revert(abi.decode(reason, (string)));
+        }
+        return reason;
+    }
+
+    function _handleRevertExactInputSingle(bytes memory reason, PoolKey memory poolKey)
+        private
+        view
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksCrossed)
+    {
+        int24 tickBefore;
+        int24 tickAfter;
+        BalanceDelta deltas;
+        deltaAmounts = new int128[](2);
+        (, tickBefore,,) = poolManager.getSlot0(poolKey.toId());
+        reason = validateRevertReason(reason);
+        (deltas, sqrtPriceX96After, tickAfter) = abi.decode(reason, (BalanceDelta, uint160, int24));
+        deltaAmounts[0] = deltas.amount0();
+        deltaAmounts[1] = deltas.amount1();
+
+        initializedTicksCrossed =
+            PoolTicksCounter.countInitializedTicksCrossed(poolManager, poolKey, tickBefore, tickAfter);
```

Comment:
why do we do the counting _post revert_ here but _pre revert_ for exactInput? think you could refactor this into `_quoteExactInputSingle` to deduplicate code across the paths